py_ecc/fields/optimized_field_elements.py

```diff
@@ -0,0 +1,395 @@
+        if isinstance(other, FQ):
+            return self.n == other.n
+        elif isinstance(other, int):
+            return self.n == other
+        else:
+            raise ValueError(
+                "Expected an int or FQ object, but got object of type {}"
+                .format(type(other))
+            )
+
+    def __ne__(self, other: IntOrFQ) -> bool:    # type:ignore # https://github.com/python/mypy/issues/2783 # noqa: E501
+        return not self == other
+
+    def __neg__(self) -> "FQ":
+        return FQ(-self.n, self.curve_name)
+
+    def __repr__(self) -> str:
+        return repr(self.n)
+
+    def __int__(self) -> int:
+        return self.n
+
+    @classmethod
+    def one(cls, curve_name: str) -> "FQ":
+        return cls(1, curve_name)
+
+    @classmethod
+    def zero(cls, curve_name: str) -> "FQ":
+        return cls(0, curve_name)
+
+
+class FQP(object):
+    """
+    A class for elements in polynomial extension fields
+    """
+    degree = 0  # type: int
+    mc_tuples = None  # type: List[Tuple[int, int]]
+    curve_name = None
+    field_modulus = None
+
+    def __init__(self,
+                 coeffs: Sequence[IntOrFQ],
+                 curve_name: str,
+                 modulus_coeffs: Sequence[IntOrFQ]=None) -> None:
+        """
+        curve_name can be either 'bn128' or 'bls12_381'
+        This is needed to obtain field_modulus, FQ2_MODULUS_COEFFS
+        and FQ12_MODULUS_COEFFS from the curve properties
+        """
+        self.curve_name = curve_name
+        self.field_modulus = field_properties[curve_name]["field_modulus"]
+
+        if len(coeffs) != len(modulus_coeffs):
+            raise Exception(
+                "coeffs and modulus_coeffs aren't of the same length"
+            )
+        self.coeffs = tuple(coeffs)
+        # The coefficients of the modulus, without the leading [1]
+        self.modulus_coeffs = tuple(modulus_coeffs)
+        # The degree of the extension field
+        self.degree = len(self.modulus_coeffs)
+
+    def __add__(self, other: "FQP") -> "FQP":
+        if not isinstance(other, type(self)):
+            raise ValueError(
+                "Expected an FQP object, but got object of type {}"
+                .format(type(other))
+            )
+
+        return type(self)([
+            int(x + y) % self.field_modulus
+            for x, y
+            in zip(self.coeffs, other.coeffs)
+        ], self.curve_name)
+
+    def __sub__(self, other: "FQP") -> "FQP":
+        if not isinstance(other, type(self)):
+            raise ValueError(
+                "Expected an FQP object, but got object of type {}"
+                .format(type(other))
+            )
+
+        return type(self)([
+            int(x - y) % self.field_modulus
+            for x, y
+            in zip(self.coeffs, other.coeffs)
+        ], self.curve_name)
+
+    def __mod__(self, other: Union[int, "FQP"]) -> "FQP":
+        return self.__mod__(other)
```

Comment:
What the heck is this? :eyes: :fire:  It just recurses until it explodes on stack depth.

I mean, I see it's not new code: it was in the old implementation. It just doesn't make sense, and clearly must not be getting tested.

Maybe we should just raise a `NotImplementedError` until we can figure out what should be happening here. (Note that there are a couple of these)